### PR TITLE
[llvm-profgen] Add --sample-period to estimate absolute counts

### PR DESCRIPTION
**Pull request by tcreech-intel**
Without `--sample-period`, no assumptions are made about perf profile sample frequencies. This is useful for comparing relative hotness of different program locations within the same profile.

With `--sample-period`, LBR- and IP-based profile hit counts are adjusted to estimate the absolute total event count for each program location. This makes it reasonable to compare hit counts between different profiles, e.g., between two LBR-based execution frequency profiles with different sampling periods or between LBR-based execution frequency profiles and IP-based branch mispredict profiles.

This functionality is in support of HWPGO[^1], which aims to enable feedback from a wider range of hardware events.

[^1]: https://llvm.org/devmtg/2024-04/slides/TechnicalTalks/Xiao-EnablingHW-BasedPGO.pdf

### CODE REVIEW

**Comment by tcreech-intel** [comment]
@williamweixiao, @HaohaiWen, it may also be useful to have this PR, which adds `--sample-period` functionality. This is a nice usability improvement.
(It's a separate PR because it depends on the changes in #99026.)

**Review by williamweixiao — APPROVED** [review]
We definitely need this feature when handing multiple events together.

File: llvm/tools/llvm-profgen/PerfReader.cpp [hunk 1]
```diff
@@ -53,6 +53,10 @@ static cl::alias
                           cl::desc("Comma-delimited version of -perf-event"),
                           cl::aliasopt(PerfEventFilter));
 
+static cl::opt<uint64_t>
+    SamplePeriod("sample-period", cl::init(1),
```
Comment (WenleiHe) [inline llvm/tools/llvm-profgen/PerfReader.cpp:57]: This can be confusing -- having default value 1 implies we always assume a sample period of 1.

**Comment by WenleiHe** [comment]
I don't get it why --sample-period is needed from the change summary. 

It sounds like the rational behind this change is two-fold:

1) To normalize LBR sample with IP sample. If that's the case, we only need to normalize by LBR depth regardless of sampling rate/period. 

2) To support calibration and merging of profile collected with different sampling rate/period. If that's the case, llvm-`profdata merge -weighted-input ..` is the way to go. 

Let me know if I misunderstood the intention of the change, otherwise I'd suggest not taking this patch. We need to be careful about avoiding duplicating/overlapping features, avoiding proliferation of switches, for long term maintainability.

**Comment by tcreech-intel** [comment]
I'm interested in handling the combination of 1) and 2), where we're sampling the IP-only profile at a different period from the LBR-based profile and want to compare the two, e.g., for computing branch miss ratios. 

However, let's say we're only concerned with 1), i.e., dividing LBR hit counts by the LBR depth, because we can account for sampling periods with `llvm-profdata` post-processing.  If we apply the sampling period then we can do this division while processing each sample's contribution. This is useful because the effective LBR depth can vary by platform and even across samples in the same profile. Without applying the sampling period first, `Count` will typically be smaller than `Sample->LBRStack.size() - 1` and the result won't be useful.

In short, --sample-period lets us avoid dividing integers which are smaller than LBR depths without assuming anything about LBR depths. (It also produces profile counts which approximate absolute event counts, but we don't depend on that property.)

Thoughts here are appreciated, but I will move to revert these two llvm-profgen changes.

**Comment by WenleiHe** [comment]
> This is useful because the effective LBR depth can vary by platform and even across samples in the same profile. 

But when processing profile, we know the exact depth of each LBR sample?

> Without applying the sampling period first, Count will typically be smaller than Sample->LBRStack.size() - 1 and the result won't be useful.

Ok, I see why you want to multiple the count with something first. In this case, the multiplier doesn't have to be sampling period, instead it just needs to be something bigger than max LBR depth? It's sort of like a normalizing factor.. So I think we can get away with something simpler than actual sample period -- like we don't want to suggest that user need to provide the actual sampling period.

**Comment by tcreech-intel** [comment]
> But when processing profile, we know the exact depth of each LBR sample?

Yes, this should be `Sample->LBRStack.size()`.

> In this case, the multiplier doesn't have to be sampling period, instead it just needs to be something bigger than max LBR depth? It's sort of like a normalizing factor.. So I think we can get away with something simpler than actual sample period

Yes, for purposes of normalizing LBR and IP-only profiles with uniform sampling period I think that's true. It doesn't allow a user to compare profiles of different sampling periods, though, and I think that will be a useful capability. (Both perf and SEP support per-event periods.)

With a fixed normalizing factor users in that case would need to use their sampling periods to come up with an adjustment factor and apply it with `llvm-profdata merge`. I certainly understand the appeal of organizing the tooling this way, but I do worry some users will have trouble with this extra step.

> we don't want to suggest that user need to provide the actual sampling period

Stepping back a bit, another direction might be to have `llvm-profgen` run `perf script` asking for the "`period`" field, i.e., to include the sampling period associated with each sample. (In general perf allows for the period to vary within a profile.) This could be optional behavior. Do you think that might be reasonable?
